### ISSUE TITLE
Display Internal FPS (Frame Duping method) is inaccurate

### ISSUE THREAD

**Issue by ggdrt** [libretro/beetle-psx-libretro#633]
The way Internal FPS metric is calculated is based on the number of GP1(05h) calls:
https://github.com/libretro/beetle-psx-libretro/blob/c7c8ffd3d2cf1c780ccf98417e60908f9adf970d/mednafen/psx/gpu.cpp#L1166-L1171

This GPU command sets the upper left coordinate of the scanout display area (or framebuffer, depending on what terminology you want to use). Most games call this command on every new frame that it wants to output. 

However, since the coordinates are latched in the GPU and video output will be continually generated from the display area as long as the display is on, it is also perfectly valid for a game to continue drawing to the current display area and expect the new things it has drawn to show up in the video output. For example, Final Fantasy VII calls this GPU command every four frames in battle sequences (~14.96 reported Internal FPS), but menu and UI elements (e.g. Tifa's Limit Break slots) animate at half or even full refresh rate because the game continually draws to the same display area before calling the command to switch to the next display area.

The frame duping core option is based on the Internal FPS metric, meaning that in some cases with frame duping enabled we could be dropping frames where we shouldn't be.

For improvements, we could check if there were any draws to the display area before doing a video callback to improve the Internal FPS metric, but that far outweighs any performance benefits we get from doing frame duping. I'm not sure if there's a practical solution to this, maybe I'm missing something.

**Comment by simias** [comment]
>For improvements, we could check if there were any draws to the display area before doing a video callback to improve the Internal FPS metric, but that's costly in the general case and far outweighs any performance benefits we get from doing frame duping

We could only enable that code when the clipping area overlaps the display area (which I'm guessing should be quite rare, although maybe I'd be surprised). If the buffer being displayed is completely outside of the draw area there's no need to test anything. That would mean hooking into GP0(E3h) and GP0(E4) calls to update a boolean variable saying if an overlap is even possible.

Beyond that I'm not sure what would be the performance impact of checking for draws to the display area, after all it's just a bounding box test which is reasonably cheap. We could even be a bit messier and just test one vertex of every tri/quad and it'd still probably good enough to handle most of the cases (including your FF menu scenario).

The current implementation of the internal FPS counter is definitely hacky, I did it originally to test the overclock code, I didn't expect that it would end up being used to drive other features. I guess it works well enough for most games.

**Comment by ggdrt** [comment]
Ah shoot, knew I was missing something. Forgot about the draw area clipping commands. I was overthinking a scenario where we would have to save every vertex of every rendered line/poly/rect and test them against the display area at scanout time.

Thanks for suggestions, I'll give them a shot and see how far the feature can be improved.

**Comment by rapierx** [comment]
> The frame duping core option is based on the Internal FPS metric, meaning that in some cases with frame duping enabled we could be dropping frames where we shouldn't be.

Correct. With frame duping ON, loading screens are dropped on Twisted Metal 4. Loading screen is replaced with a black frame or loading bars doesn't refresh.

Frame duping ON, loading bar barely refreshes:

https://streamable.com/ndl8s

Frame duping OFF, loading bar OK:

https://streamable.com/dmqm5

**Comment by simias** [comment]
There are a few commands that can draw to the screen regardless of the draw area. Out of the top of my head:

* `Command_FBWrite`: normally used to load textures (through DMA or CPU writes), can write to the display area directly (might be used for FMVs or static loading screens in some games?)

* `Command_FBFill`: used to fill an area with a solid color

* `Command_FBCopy`: copies one rectangle in VRAM

I don't know if it's worth hooking into these functions to detect writes to the display area but it might be worth keeping in mind if more problematic games crop up later.

**Comment by ggdrt** [comment]
Turns out the solution to this problem, at least for render commands, was fairly trivial -- we just need to check if drawing to the display area was enabled for any GP0(E1h) commands that particular frame. I think it's possible to get false positives from this method but it's certainly better than false negatives for proper frame duping. Credits to @stenzek's handy debugger in duckstation for helping figure this one out.


Hooking into the VRAM transfer commands might be necessary after all. FF7's battle swirl doesn't work correctly otherwise.

**Comment by simias** [comment]
As far as I know there's not really any drawback for a game to leave the "draw to display area" option enabled even if they don't actually want to do it, so I wouldn't be too surprised if some game just set it, even if by mistake. I guess it's worth trying.

**Comment by LibretroAdmin** [comment]
Implemented this now
https://github.com/libretro/beetle-psx-libretro/commit/e46f509f5d99f18dc4bf19e974d2caffd94ffc7a

Will need further testing
